### PR TITLE
Add automatic PyTorch library discovery for Python applications

### PR DESCRIPTION
**Pull request by mradosav-amd**
## Motivation

When profiling PyTorch applications with rocprof-sys-run or rocprof-sys-sample, the PyTorch libraries may not be in LD_LIBRARY_PATH, causing runtime linking issues. This change automatically discovers and adds the PyTorch library path when profiling Python applications that use PyTorch.

## Technical Details

- Added `is_python_interpreter()` function to detect if the target executable is a Python interpreter (python, python3, python3.X)
- Added `discover_torch_libpath()` function that runs Python to query torch installation path via `torch.__path__`
- Added `add_torch_library_path()` template function to prepend discovered torch lib directory to LD_LIBRARY_PATH
- Added `consolidate_env_entries()` function to deduplicate environment variable entries (removes duplicate paths in colon-separated values)
- Integrated torch library path discovery into both `rocprof-sys-run` and `rocprof-sys-sample` tools

Known issue:
- Unable to detect when to add LD_LIBRARY_PATH if Python is run via a shell script, for example

## JIRA ID

AIPROFSYST-107

## Test Plan

- Unit tests for `is_python_interpreter()` covering various Python interpreter names and paths
- Unit tests for `consolidate_env_entries()` verifying deduplication of environment variable entries
- Manual testing: Profile a PyTorch Python application with rocprof-sys-run and verify torch libraries are loaded correctly

## Test Result

- `is_python_interpreter()` should correctly identify python, python3, python3.X executables (including full paths) and reject non-Python executables
- `consolidate_env_entries()` should merge duplicate environment variable keys and deduplicate path entries
- PyTorch applications should run successfully under profiling without manual LD_LIBRARY_PATH configuration
## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-systems/source/lib/common/tests/test_environment.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/tests/test_environment.cpp]: The year needed update, but you can use the short header instead.

```suggestion
// Copyright (c) Advanced Micro Devices, Inc.
// SPDX-License-Identifier:  MIT
```

File: projects/rocprofiler-systems/source/lib/common/tests/test_environment.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    };
+
+    consolidate_env_entries(env_vars);
+
+    ASSERT_EQ(env_vars.size(), 1);
+    EXPECT_STREQ(env_vars[0], "PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/bin2");
+
+    for(auto* entry : env_vars)
+        free(entry);
+}
```
Comment (dgaliffiAMD) [inline projects/rocprofiler-systems/source/lib/common/tests/test_environment.cpp:63]: Suggesting some additional edge-case checks. 

```suggestion
}

TEST_F(DuplicatedEnvironmentEntriesTest, HandlesEmptyVector)
{
    std::vector<char*> env_vars;
    consolidate_env_entries(env_vars);
    EXPECT_TRUE(env_vars.empty());
}

TEST_F(DuplicatedEnvironmentEntriesTest, HandlesNullEntries)
{
    std::vector<char*> env_vars = {
        strdup("PATH=/usr/bin"),
        nullptr,
        strdup("PATH=/bin"),
    };

    consolidate_env_entries(env_vars);

    ASSERT_EQ(env_vars.size(), 1);
    EXPECT_STREQ(env_vars[0], "PATH=/usr/bin:/bin");

    for(auto* entry : env_vars)
        std::free(entry);
}

TEST_F(DuplicatedEnvironmentEntriesTest, HandlesEmptyValues)
{
    std::vector<char*> env_vars = {
        strdup("EMPTY_VAR="),
        strdup("PATH=/usr/bin"),
    };

    consolidate_env_entries(env_vars);

    ASSERT_EQ(env_vars.size(), 2);
    
    for(auto* entry : env_vars)
        std::free(entry);
}

class AddTorchLibraryPathTest : public ::testing::Test
{
protected:
    std::unordered_set<std::string> updated_envs;
};

TEST_F(AddTorchLibraryPathTest, SkipsNonPythonExecutables)
{
    std::vector<char*> envp = {
        strdup("LD_LIBRARY_PATH=/usr/lib"),
    };
    std::vector<char*> argv = {
        strdup("/usr/bin/bash"),
    };

    add_torch_library_path(envp, argv, false, updated_envs);

    // Should not modify environment
    ASSERT_EQ(envp.size(), 1);
    EXPECT_STREQ(envp[0], "LD_LIBRARY_PATH=/usr/lib");

    for(auto* entry : envp) std::free(entry);
    for(auto* entry : argv) std::free(entry);
}

TEST_F(AddTorchLibraryPathTest, HandlesEmptyArgv)
{
    std::vector<char*> envp = {
        strdup("LD_LIBRARY_PATH=/usr/lib"),
    };
    std::vector<char*> argv;

    add_torch_library_path(envp, argv, false, updated_envs);

    ASSERT_EQ(envp.size(), 1);
    EXPECT_STREQ(envp[0], "LD_LIBRARY_PATH=/usr/lib");

    for(auto* entry : envp) std::free(entry);
}

TEST_F(AddTorchLibraryPathTest, HandlesNullArgvFront)
{
    std::vector<char*> envp = {
        strdup("LD_LIBRARY_PATH=/usr/lib"),
    };
    std::vector<char*> argv = { nullptr };

    add_torch_library_path(envp, argv, false, updated_envs);

    ASSERT_EQ(envp.size(), 1);
    for(auto* entry : envp) std::free(entry);
}
```

**Review by dgaliffiAMD — APPROVED** [review]

File: projects/rocprofiler-systems/source/lib/common/environment.hpp [hunk 3]
```diff
@@ -335,13 +445,153 @@ update_env(std::vector<char*>& _environ, std::string_view _env_var, Tp&& _env_va
...
+        auto eq_pos = entry.find('=');
+        if(eq_pos == std::string_view::npos) return std::nullopt;
+        return std::make_pair(entry.substr(0, eq_pos), entry.substr(eq_pos + 1));
+    };
+
+    auto join_parts = [delim](std::string_view                key,
+                              const std::vector<std::string>& parts) {
+        std::string result;
+
+        const auto total_parts_length = std::accumulate(
```
Comment (habajpai-amd) [inline projects/rocprofiler-systems/source/lib/common/environment.hpp:525]: The return type of std::accumulate is determined by initial value's type, not the lambda's return type.

So even though the lambda returns std::size_t, the accumulator is int. This causes overflow or truncation on 64-bit.

Fix
const auto total_parts_length = std::accumulate(
    parts.begin(), parts.end(), std::size_t{0},  // <- correct type
    [](std::size_t acc, const std::string& part) { return acc + part.size(); });

Comment (mradosav-amd) [reply projects/rocprofiler-systems/source/lib/common/environment.hpp:525]: good catch. thanks.
